File: app/src/main/java/com/github/android/lvrn/lvrnproject/view/dialog/notebookselection/impl/NotebookSelectionPresenterImpl.java
package com.github.android.lvrn.lvrnproject.view.dialog.notebookselection.impl;

import android.support.v7.widget.RecyclerView;

import com.github.valhallalabs.laverna.persistent.entity.Notebook;
import com.github.android.lvrn.lvrnproject.service.core.NotebookService;
import com.github.android.lvrn.lvrnproject.util.PaginationArgs;
import com.github.android.lvrn.lvrnproject.view.dialog.notebookselection.NotebookSelectionDialogFragment;
import com.github.android.lvrn.lvrnproject.view.dialog.notebookselection.NotebookSelectionPresenter;
import com.github.android.lvrn.lvrnproject.view.listener.RecyclerViewOnScrollListener;
import com.github.android.lvrn.lvrnproject.util.CurrentState;

import java.util.List;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import io.reactivex.subjects.ReplaySubject;

/**
 * @author Vadim Boitsov <vadimboitsov1@gmail.com>
 */

class NotebookSelectionPresenterImpl implements NotebookSelectionPresenter {

    private NotebookSelectionDialogFragment mNotebookSelectionDialogFragment;

    private NotebookService mNotebookService;

    private Disposable mPaginationDisposable;

    private List<Notebook> mNotebooks;

    NotebookSelectionPresenterImpl(NotebookService mNotebookService) {
        (this.mNotebookService = mNotebookService).openConnection();
    }

    @Override
    public void bindView(NotebookSelectionDialogFragment notebookSelectionDialogFragment) {
        mNotebookSelectionDialogFragment = notebookSelectionDialogFragment;
        if (!mNotebookService.isConnectionOpened()) {
            mNotebookService.openConnection();
        }
    }

    @Override
    public void unbindView() {
        mNotebookService.closeConnection();
    }

    @Override
    public void subscribeRecyclerViewForPagination(RecyclerView recyclerView) {
        ReplaySubject<PaginationArgs> paginationReplaySubject;

        mPaginationDisposable = (paginationReplaySubject = ReplaySubject.create())
                .observeOn(Schedulers.io())
                .map(paginationArgs -> mNotebookService.getByProfile(CurrentState.profileId, paginationArgs))
                .filter(notebooks -> !notebooks.isEmpty())
                .map(newNotebooks -> mNotebooks.addAll(newNotebooks))
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(aBoolean -> mNotebookSelectionDialogFragment.updateRecyclerView(),
                        throwable -> {/*TODO: find out what can happen here*/});

        recyclerView.addOnScrollListener(new RecyclerViewOnScrollListener(paginationReplaySubject));
    }

    @Override
    public void disposePagination() {
        if( mPaginationDisposable != null && !mPaginationDisposable.isDisposed()) {
            mPaginationDisposable.dispose();
        }
    }

    @Override
    public List<Notebook> getNotebooksForAdapter() {
        mNotebooks = mNotebookService.getByProfile(CurrentState.profileId, new PaginationArgs());
        return mNotebooks;
    }
}

File: app/src/main/java/com/github/android/lvrn/lvrnproject/view/dialog/notebookselection/impl/NotebookSelectionPresenterImpl.kt
package com.github.android.lvrn.lvrnproject.view.dialog.notebookselection.impl

import androidx.recyclerview.widget.RecyclerView
import com.github.android.lvrn.lvrnproject.service.core.NotebookService
import com.github.android.lvrn.lvrnproject.util.CurrentState.Companion.profileId
import com.github.android.lvrn.lvrnproject.util.PaginationArgs
import com.github.android.lvrn.lvrnproject.view.dialog.notebookselection.NotebookSelectionDialogFragment
import com.github.android.lvrn.lvrnproject.view.dialog.notebookselection.NotebookSelectionPresenter
import com.github.android.lvrn.lvrnproject.view.listener.RecyclerViewOnScrollListener
import com.github.valhallalabs.laverna.persistent.entity.Notebook
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.functions.Function
import io.reactivex.schedulers.Schedulers
import io.reactivex.subjects.ReplaySubject
import javax.inject.Inject

/**
 * @author Vadim Boitsov <vadimboitsov1></vadimboitsov1>@gmail.com>
 */
internal class NotebookSelectionPresenterImpl @Inject constructor(val mNotebookService: NotebookService) :
    NotebookSelectionPresenter {
    private var mNotebookSelectionDialogFragment: NotebookSelectionDialogFragment? = null

    private var mPaginationDisposable: Disposable? = null

    private var mNotebooks: MutableList<Notebook> = mutableListOf()

    init {
        mNotebookService.openConnection()
    }

    override fun bindView(notebookSelectionDialogFragment: NotebookSelectionDialogFragment?) {
        mNotebookSelectionDialogFragment = notebookSelectionDialogFragment
        if (!mNotebookService.isConnectionOpened()) {
            mNotebookService.openConnection()
        }
    }

    override fun unbindView() {
        mNotebookService.closeConnection()
    }

    override fun subscribeRecyclerViewForPagination(recyclerView: RecyclerView) {
        var paginationReplaySubject: ReplaySubject<PaginationArgs>

        mPaginationDisposable =
            ReplaySubject.create<PaginationArgs>().also { paginationReplaySubject = it }
                .observeOn(Schedulers.io())
                .map { paginationArgs: PaginationArgs? ->
                    mNotebookService.getByProfile(
                        profileId!!, paginationArgs!!
                    )
                }
                .filter { notebooks: List<Notebook> -> !notebooks.isEmpty() }
                .map { newNotebooks: List<Notebook> ->
                    mNotebooks.addAll(
                        newNotebooks
                    )
                }
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                    { aBoolean: Boolean? -> mNotebookSelectionDialogFragment!!.updateRecyclerView() },
                    { throwable: Throwable? -> })

        recyclerView.addOnScrollListener(RecyclerViewOnScrollListener(paginationReplaySubject))
    }

    override fun disposePagination() {
        if (mPaginationDisposable != null && !mPaginationDisposable!!.isDisposed) {
            mPaginationDisposable!!.dispose()
        }
    }

    override val notebooksForAdapter: List<Notebook>
        get() {
            mNotebooks = mNotebookService.getByProfile(profileId!!, PaginationArgs()).toMutableList()
            return mNotebooks
        }
}
